src/mono/netcore/System.Private.CoreLib/src/System/Reflection/Emit/AQNBuilder.Mono.cs

```diff
@@ -0,0 +1,483 @@
+			bool m_bNestedName;
+			bool m_bHasAssemblySpec;
+			bool m_bUseAngleBracketsForGenerics;
+			Stack m_stack = new Stack ();
+
+			public void Dispose () { m_pStr = null; }
+			public void OpenGenericArguments()
+			{
+				if (!CheckParseState(ParseState.NAME))
+					Fail();
+
+				m_parseState = ParseState.START;
+				m_instNesting ++;
+				m_bFirstInstArg = true;
+
+				if (m_bUseAngleBracketsForGenerics)
+					Append('<');
+				else
+					Append('[');
+
+			}
+			public void CloseGenericArguments()
+			{
+				if (m_instNesting == 0)
+					Fail();
+				if (!CheckParseState(ParseState.START))
+					Fail();
+
+				m_parseState = ParseState.GENARGS;
+
+				m_instNesting --;
+
+				if (m_bFirstInstArg)
+				{
+					m_pStr.Remove(m_pStr.Length - 1, 1);
+				}
+				else
+				{
+					if (m_bUseAngleBracketsForGenerics)
+						Append('>');
+					else
+						Append(']');
+				}
+			}
+			public void OpenGenericArgument()
+			{
+				if(!CheckParseState (ParseState.START))
+					Fail ();
+
+				if (m_instNesting == 0)
+					Fail ();
+
+				m_parseState = ParseState.START;
+				m_bNestedName = false;
+
+				if (!m_bFirstInstArg)
+					Append(',');
+
+				m_bFirstInstArg = false;
+
+				if (m_bUseAngleBracketsForGenerics)
+					Append('<');
+				else
+					Append('[');
+				PushOpenGenericArgument();
+			}
+
+			public void CloseGenericArgument()
+			{
+				if (!CheckParseState(ParseState.NAME | ParseState.GENARGS | ParseState.PTRARR | ParseState.BYREF | ParseState.ASSEMSPEC))
+					Fail();
+
+				if (m_instNesting == 0)
+					Fail();
+
+				m_parseState = ParseState.START;
+
+				if (m_bHasAssemblySpec)
+				{
+					if (m_bUseAngleBracketsForGenerics)
+						Append('>');
+					else
+						Append(']');
+				}
+
+				PopOpenGenericArgument();
+			}
+			public void AddName(string szName)
+			{
+				if (szName == null)
+					Fail ();
+
+				if (!CheckParseState(ParseState.START | ParseState.NAME))
+					Fail ();
+
+				m_parseState = ParseState.NAME;
+
+				if (m_bNestedName)
+					Append ('+');
+
+				m_bNestedName = true;
+
+				EscapeName (szName);
+			}
+
+			public void AddPointer()
+			{
+				if (!CheckParseState(ParseState.NAME | ParseState.GENARGS | ParseState.PTRARR))
+					Fail ();
+
+				m_parseState = ParseState.PTRARR;
+
+				Append ('*');
+			}
+
+			public void AddByRef()
+			{
+				if (!CheckParseState(ParseState.NAME | ParseState.GENARGS | ParseState.PTRARR))
+					Fail();
+
+				m_parseState = ParseState.BYREF;
+
+				Append('&');
+			}
+
+			public void AddSzArray()
+			{
+				if (!CheckParseState(ParseState.NAME | ParseState.GENARGS | ParseState.PTRARR))
+					Fail();
+
+				m_parseState = ParseState.PTRARR;
+
+				Append("[]");
+			}
+			public void AddArray(int rank)
+			{
+				if (!CheckParseState(ParseState.NAME | ParseState.GENARGS | ParseState.PTRARR))
+					Fail();
+
+				m_parseState = ParseState.PTRARR;
+
+				if (rank <= 0)
+					throw new ArgumentOutOfRangeException ();
+
+				if (rank == 1)
+					Append("[*]");
```

Comment:
Nit: we try to use braces on all blocks in a cascading if/else if any of them use braces